Several changes

chris-ha458 opened the pull request:
I know this PR combines a lot into one,  but they are mostly subsurface changes that don't touch each other.
Also they each incorporate their own doctests to check for correctness, but we could always add more.

1. I added a Ranked-unionfind implementation. The previous one only uses path compression, which is still good, but adding rank structure further improves it. The only way to improve it from here is to find and apply a concurrent way to implement it.
2. I added ftfy. Nvidia Curator uses this seemingly before processing. This only adds the utility functions. when and where to deploy it requires further discussion.
3. Shardify remaining batch structures. Using 'datasets' native sharding is superior to batching, so i replaced all instances of it.

ChenghaoMou commented on text_dedup/utils/ftfy_utils.py:
```diff
@@ -0,0 +1,103 @@
+from ftfy import fix_text
+
+
+def fix_text_batch(batch: Dict[str, List], column: str = "text") -> Dict[str, List]:
```
Here is one concern I have:

In this function, the text column is overwritten, which is fine if you want to preprocess your data beforehand, but again that's outside the scope of this project.

But if this is intended for preparing for deduplication, then you don't really need the batch function(s). You can simply add the fixing in the fingerprinting function/stage to maximize CPU usage.

chris-ha458 replied:
It could be either or both.
I did intend for both but I am still not sure the best way to implement it in each case (it's also different for exact and near dedup)
As you can see, these are currently added as helpers to lay the groundwork ATM.

chris-ha458 replied:
Hmmm so if I understand you correctly it is not useful for fingerprinting and for this project isn't dealing preprocessing that the batches aren't necessary in any case?
If that is so I'm happy to remove the batching functions.

ChenghaoMou replied:
I am fine with `ftfy` being introduced. We can introduce the fixing on the fly during the fingerprinting. For example:

```
tokens: Set[bytes] = {
        bytes(" ".join(t).lower(), "utf-8") for t in ngrams(NON_ALPHA.split(fix_text(content)), ngram_size, min_length)
    }
```

This keeps all the CPU computation together in one go and leaves no storage/cache footprint, so we don't have to iterate the dataset or create a new text column beforehand. Another point is that this repo shouldn't change the original data but only remove deduplicates.

However, this also leads to the question of how the fixing behaviour is controlled, with so many options `ftfy` offers and settings each script already has. It is probably better to offer a config parsing to digest all the settings in the long run.

ChenghaoMou reviewed (commented):
Thanks for the PR!

chris-ha458 commented:
added Documentation regarding #26 and #50. I think the a more robust solution involving `mp.shared_memory()` or `mp.manager()` could be possible, but that is way beyond the scope of this PR.

chris-ha458 commented:
As for ftfy related, I think this PR is getting bogged down by it so I removed most of it.
They werent really necessary since 1. since this repo shouldn't change the original data. and 2. Batching the way i implemented it didn't show performance improvements in my experiments.

I left the pyproject.toml changes and ftfy_utils.py to allow for iterations and experiments with minimal input.

As for your suggestion code :

```tokens: Set[bytes] = {
        bytes(" ".join(t).lower(), "utf-8") for t in ngrams(NON_ALPHA.split(fix_text(content)), ngram_size, min_length)
    }```

That is looks close to what I had in mind but i was also thinking about something like this as well

```tokens: Set[bytes] = {
        bytes(fix_text(" ".join(t).lower(), "utf-8") for t in ngrams(NON_ALPHA.split(content), ngram_size, min_length))
    }```

There's overt and subtle differences in both behavior and performance and I will try to address and document them in another PR.
I am also preparing another performance related PR on top of this one that touches on specifically that part of the code so I am hesitant to implement something as part of this PR.

Let me know any other blockers regarding this PR. When it gets merged, i will try to explore some more options inlcuding the ones already discussed and implement or document them as applicable.

ChenghaoMou commented on text_dedup/utils/ftfy_utils.py:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# @Date    : 2022-12-26 15:37:44
```
Nitpicking here: the date is wrong.

ChenghaoMou reviewed (approved):
LGTM
